### PR TITLE
ARROW-14053: [C++][CSV] Use atomic counter for async tests

### PR DESCRIPTION
**Pull request by n3world**

### CODE REVIEW

File: cpp/src/arrow/csv/reader_test.cc [hunk 1]
```diff
@@ -176,20 +177,21 @@ void TestNestedParallelism(std::shared_ptr<internal::ThreadPool> thread_pool,
...
+  std::atomic<int> num_invalid_rows(0);
   opts.invalid_row_handler = [&](const InvalidRow& row) {
     ++num_invalid_rows;
-    if (rows_counted) {
+    if (async) {
+      // Row numbers are not counted in batches during async processing
+      EXPECT_EQ(-1, row.number);
+    } else {
       // actual data starts at row #2 in the CSV "file"
       EXPECT_EQ(num_invalid_rows * INVALID_EVERY + 1, row.number);
```
Comment (pitrou) [inline cpp/src/arrow/csv/reader_test.cc:194]: Is `num_invalid_rows` still unchanged from above?

**Comment by pitrou** [comment]
@github-actions crossbow submit test-ubuntu-20.04-cpp-thread-sanitizer

File: cpp/src/arrow/csv/reader_test.cc (hunk 1 above)
Comment (n3world) [reply cpp/src/arrow/csv/reader_test.cc:194]: Yes. When it is not async the blocks are processed serially and so is the callback. I can store the value locally and do a postfix increment just for atomic consistency.

**Review by pitrou — APPROVED** [review]
+1, thank you. I will just wait for CI and then merge if everything's ok.

**Review by kou — APPROVED** [review]
+1

**Pull request closed without merging**